driver/reducecalculation_driver.hpp

```diff
@@ -350,32 +425,64 @@ Tref ReduceCalculationDriver<Tgpu, Tref>::GetTolerance()
     return tolerance;
 }
 
-template <typename Tgpu, typename Tref>
-int ReduceCalculationDriver<Tgpu, Tref>::VerifyForward()
+template <typename Tgpu, typename Tref, typename Tgpu_out>
+int ReduceCalculationDriver<Tgpu, Tref, Tgpu_out>::NumericalVerifyForward()
+{
+}
+
+template <typename Tgpu, typename Tref, typename Tgpu_out>
+int ReduceCalculationDriver<Tgpu, Tref, Tgpu_out>::VerifyForward()
 {
     RunForwardCPU();
-    const Tref tolerance = GetTolerance();
-    auto error           = miopen::rms_range(outhost, out);
 
-    if(!std::isfinite(error) || error > tolerance)
+    if(isLogicalCalculation)
     {
-        std::cout << "Forward Reduce Calculation FAILED: " << error << " > " << tolerance
-                  << std::endl;
-        return EC_VerifyFwd;
+        if constexpr(std::is_same<Tgpu_out, Tref>::value)
+        {
+            auto is_equal = (outhost == out);
+
+            if(!is_equal)
+            {
+                std::cout << "Forward Reduce Logical Calculation FAILED: " << std::endl;
+                return EC_VerifyFwd;
+            }
+            else
+            {
+                std::cout << "Forward Reduce Logical Calculation Verifies OK on CPU reference"
+                          << std::endl;
+            }
+        }
+        else
+        {
+            std::cout << "Forward Reduce Logical Calculation FAILED: Type mismatch: Cannot compare "
+                         "outhost and out"
+                      << std::endl;
```

Comment:
you setup yourself Tgpu_out and Tref, if its different, it can not compare so why you separate it? redundant if else

Reply:
I removed `Tgpu_out`, pls check my  new solution